calculation_scripts/OptimizationLoop.py

```diff
@@ -0,0 +1,167 @@
+    light_direction = [0, 0, 1]
+    light_polarization = [1, 0, 0]
+    light_amplitude = 1.0
+    light_wavelength_nm = 542
+    background_refractive_index = np.sqrt(
+        np.real(dielectric_materials[0]))
+
+    # Numerical optimization parameters.
+    # A-product-core periodicity parameters / boundaries.
+    apc_max_m = 50
+    apc_max_n = 50
+    apc_l_m = 22
+    apc_l_n = 22
+    apc_method_name = "FCD"
+
+    # Cached integral values.
+    sici_delta = 0.1
+    sici_n = 1_000_000
+    integral_positions = np.linspace(sici_delta, sici_delta * sici_n, sici_n - 1)
+    si, ci = sici(integral_positions)
+
+    ordered_parameters = {
+        # Filter parameters.
+        "filter_beta_min": filter_beta_min,
+        "filter_beta_max": filter_beta_max,
+        "filter_ita": filter_ita,
+        "filter_method": filter_method,
+        "filter_iterations": filter_iterations,
+        "filter_radii": filter_radii,
+        "filter_enable": filter_enable,
+        # Symmetry parameters.
+        "sym_method": sym_method,
+        "sym_axis": sym_axis,
+        "sym_is_periodic": sym_is_periodic,
+        # Objective and geometry.
+        "obj_name": obj_name,
+        "obj_config": obj_config,
+        "geo_indices": geo_indices,
+        "dielectric_values": dielectric_values,
+        "geo_nx": geo_nx,
+        "geo_ny": geo_ny,
+        "geo_nz": geo_nz,
+        "geo_ntotal": geo_ntotal,
+        # Incident light wave (plus some material properties).
+        "light_direction": light_direction,
+        "light_amplitude": light_amplitude,
+        "light_polarization": light_polarization,
+        "light_wavelength_nm": light_wavelength_nm,
+        "dielectric_materials": dielectric_materials,
+        "background_refractive_index": background_refractive_index,
+        # Numerical parameters.
+        "apc_max_m": apc_max_m,
+        "apc_max_n": apc_max_n,
+        "apc_l_m": apc_l_m,
+        "apc_l_n": apc_l_n,
+        "apc_method_name": apc_method_name,
+        "geo_pixel_size_nm": geo_pixel_size_nm,
+        # Sine and cosine integral values.
+        "sineIntegralValues": si,
+        "cosineIntegralValues": ci,
+        "integralDelta": sici_delta,
+        "verbose": True,
+    }
+
+    model = dda_model.DDAModel(*list(ordered_parameters.values()))
+    return model, ordered_parameters
+
+
+
+GLOBAL_MODEL = _construct_test_model()
+
+# model, parameters = _construct_test_model()
+model, parameters = GLOBAL_MODEL
+origin_polarization = [0 + 0j] * 3 * parameters["geo_ntotal"]
+# Numerical parameters for the bicongugate gradient stabilized method. 
+bgs_max_iter = 100_000
+evo_max_iter = 400
+bgs_max_error = 1e-5
+# epsilon = 0.1
+# epsilon = 0.01 # This works better
+epsilon = 0.01
+beta1 = 0.9
+beta2 = 1 - (1 - beta1) * (1 - beta1)
+num_free_parameters = len(model.getParameters())
+all_objective_values = [0] * evo_max_iter
+
+newpath = 'E:\\Calculations\\2024April27\\HourglassUsingNewRepo\\CoreStructure\\'
```

Comment:
Nit: Good practice to hoist all "settable" variables to the top of the script or at least put them in the same place. Better practice to (1) pass them as arguments on the command line via `sys.argv` or `argparse` and (2) pack everything into a `if __name__ == '__main__'` clause. https://realpython.com/if-name-main-python/

Reply:
for 1. That sounds like a good idea for bash scripting, but I might need some help in order to build that in.
for 2. I read your link and I understand what it does, but I am unsure why we need it in our case. Is it just good coding practice to have everything in a __main__? The link says that the if clause is used to pack code that shouldnt be called when the script is used as a module, but I dont expect to be using this one as a module.

Reply:
It's generally good coding hygiene to put callables, functions, and global variables into the script and to put the actual calcuation, calling the functions, etc into the main. But for research it doesn't matter very much - just do whatever is convenient, fast, and sane.